### PR TITLE
CUDA performance optimizations

### PR DESCRIPTION
**Pull request by JohannesGaessler**
This PR adds performance optimizations for GPU accelerated token generation, mostly benefiting fast GPUs like the RTX 3090. Performance optimizations can be enabled via the options `LLAMA_CUDA_BY=2` and `LLAMA_CUDA_UNROLL=1` (make) or `LLAMA_CUDA_UNROLL=ON` (cmake) at compile time. **These options degrade performance on my GTX 1070.** Build instructions (Linux):

```
git clone https://github.com/JohannesGaessler/llama.cpp llama.cpp-johannesgaessler
cd llama.cpp-johannesgaessler                               
git fetch
git switch dfyz-xor-hack
make LLAMA_CUBLAS=1 LLAMA_CUDA_BY=2 LLAMA_CUDA_UNROLL=1
```

### Implementation details

* As suggested by @dfyz in a previous PR I have eliminated the shared memory from my CUDA kernel and am using black xor magic to sum up the partial sums at the ends. This was universally faster on my RTX 3090 and my GTX 1070. But because this breaks HIP compatibility for [this PR](https://github.com/ggerganov/llama.cpp/pull/1087) it is only being used if `GGML_USE_HIPBLAS` is not defined.
* Also suggested by @dfyz : larger blocks for the CUDA kernels by assigning two rows to each block. The option `LLAMA_CUDA_BY` sets the number of rows per block. On my RTX 3090 setting this option to 2 is faster but higher values have slightly worse performance. On my GTX 1070 a value of 2 or higher causes performance degradation.
* ~~Loop unrolling: The matrices used in llama.cpp always have the same size. So the loops used during inference can be unrolled if the compiler is told how large the matrices are. This is done via moving `ncols` from a regular argument to a template argument and adding a switch statement for the various matrix sizes (8 in total). On my RTX 3090 this is faster but on my GTX 1070 it's slower. **Enabling this option significantly increases compile time.**~~
* Larger blocks in x direction: The option `LLAMA_CUDA_BX` can be set to determine the block size in x direction. Default value is 32, 64 was faster on RTX 3090.

### Results

For the RTX 3090 I used `LLAMA_CUDA_BY=2 LLAMA_CUDA_UNROLL=1`, for the GTX 1070 I did **not** use these options.

| GPU      | Model    | ms/t master | ms/t PR | ms/t no unroll |
|----------|----------|-------------|---------|----|
| RTX 3090 | 7b q4_0  |       23.49 |   20.36 | 21.04 |
| RTX 3090 | 7b q8_0  |       24.54 |   22.05 | - |
| RTX 3090 | 13b q4_0 |       37.95 |   32.85 | 34.04 |
| RTX 3090 | 33b q4_0 |       83.15 |   69.76 | 73.46 |
| GTX 1070 | 7b q4_0  |       69.20 |   67.80 | - |
| GTX 1070 | 13b q4_0 |      141.30 |  139.39 | - |

### CODE REVIEW

**Comment by SlyEcho** [comment]
> But because this breaks HIP compatibility for https://github.com/ggerganov/llama.cpp/pull/1087 it is only being used if GGML_USE_HIPBLAS is not defined.

I would not worry about HIP right now, it is still in draft, and this is one of the reasons, it's not sure if it is fully compatible with CUDA (so far it has been).

I was able to solve the issue with this kind of change:
```diff
- __shfl_xor_sync(0xffffffff, tmp, mask, 32)
+ __shfl_xor(tmp, mask, 32)
```

Well actually, the 32 I changed to 64, etc, because of the different warp size. It would be nice if it were a define or something, actually.

**Comment by bluefireexplosion** [comment]
Excellent work! Interesting analytics:

- 7B shows a 15.3% speedup in ms/t
- 13B shows a 15.5% speedup in ms/t
- 33B shows a 19.1% speedup in ms/t

It would be interesting to see if the performance gain for 65B follows the trend of slight percentage increases as the parameter count increases.

File: CMakeLists.txt
Comment (JohannesGaessler) [inline CMakeLists.txt:70]: If not set manually by the user the value 1 seems to be converted to "ON", thus causing a compile error. If you manually set it to an integer it works. I'm a total noob when it comes to cmake, does someone know how to fix this?

Comment (SlyEcho) [reply CMakeLists.txt:70]: `option()` is boolean.
It needs to be something like:
```cmake
set(LLAMA_CUDA_BY "1" CACHE STRING "llama: y block size for dmmv CUDA kernels")
```

Comment (JohannesGaessler) [reply CMakeLists.txt:70]: That seems to have fixed it, thanks.

**Comment by ggerganov** [comment]
The unroll indeed makes the compilation super long. Not sure we want to support it.
Couldn't you achieve the same thing via:

```c++
for (int i = 0; i < n; ++i) {
...
}
```

replace with:

```c++
assert(n % 32 == 0);
for (int io = 0; io < n; io += 32) {
    #pragma unroll
    for (int i = 0; i < 32; ++i) {
    ....
    }
}
```

Or something along these lines

**Comment by JohannesGaessler** [comment]
The unroll makes the compilation longer but honestly I don't care about 2 minutes longer compilation if it means I get a few % more performance.

**Comment by JohannesGaessler** [comment]
I've added additional performance numbers to the OP. The difference from unrolling is ~5%. Keep in mind that as more performance optimizations are added this number will increase. I personally favor the current solution that just unrolls the big loop (unless there are performance differences). It's easier to maintain and for debugging purposes it's always possible to just compile without it. I can quickly test the proposed approach though.

**Comment by JohannesGaessler** [comment]
I've pushed an alternative version to [this branch](https://github.com/JohannesGaessler/llama.cpp/tree/dfyz-xor-hack-alt-unroll). The value for `istride` cannot be pushed very high; the value on the branch is the max value for 33b that is bug free. Notably unrolling the inner loop does nothing to improve performance. Unrolling both the inner and the outer loop gives you the same performance and compile time as this PR. I would argue that it's fundamentally impossible to get the performance uplift without the increase in compile time in the first place: the compiler spends that time reordering and optimizing the code which was previously impossible due to the loop.

File: CMakeLists.txt
Comment (howard0su) [inline CMakeLists.txt:70]: Can avoid introduce more and more options? instead to check the compute capability to decide if we should enable these features dynamically.

Comment (JohannesGaessler) [reply CMakeLists.txt:70]: Automating performance optimizations is something that I would like to do long-term but right now I don't think we have the data necessary to judge which options should be enabled under which circumstances.

Comment (SlyEcho) [reply CMakeLists.txt:70]: I think there should be more options. There are plenty already for runtime tuning.

**Comment by SlyEcho** [comment]
I am gonna try to bring #1087 into compatibility with this.

The shuffle function seems the biggest hurdle right now, but it doesn't [seem impossible](https://github.com/ROCm-Developer-Tools/HIP/issues/1491#issuecomment-1256171886).

File: ggml-cuda.cu
Comment (ggerganov) [inline ggml-cuda.cu:87]: Init `GGML_CUDA_DMMV_BLOCK_Y` to 1 by default

```suggestion
#define GGML_CUDA_DMMV_BLOCK_X 32 // dmmv = dequantize_mul_mat_vec
#ifndef GGML_CUDA_DMMV_BLOCK_Y
#define GGML_CUDA_DMMV_BLOCK_Y 1
#endif
```

(not tested)

Comment (JohannesGaessler) [reply ggml-cuda.cu]: This redefines the macro and overrides the option provided by the user. I added a comment instead.

Comment (JohannesGaessler) [reply ggml-cuda.cu]: Wait sorry, I did it wrong.

Comment (JohannesGaessler) [reply ggml-cuda.cu]: This works as intended.

**Comment by SlyEcho** [comment]
For AMD this is roughly 20% faster:
```cuda
#define GGML_CUDA_DMMV_BLOCK_X 64 // dmmv = dequantize_mul_mat_vec
// ...
for (int mask = GGML_CUDA_DMMV_BLOCK_X/2; mask > 0; mask >>= 1) {
    tmp += __shfl_xor_sync(0xffffffff, tmp, mask, GGML_CUDA_DMMV_BLOCK_X);
}
``` 

And `LLAMA_CUDA_BY=4 LLAMA_CUDA_UNROLL=ON` adds another couple percents. But the unrolled code is very slow to compile, especially if all architectures are enabled as they are by default.

**Comment by JohannesGaessler** [comment]
I think I've found an optimization option that's 5% faster than loop unrolling on my RTX 3090 and where loop unrolling actually degrades performance. It essentially works by processing more values per iteration in the loop instead of unrolling the loop. Implementation could be a little tricky. So you can either wait until I've worked out the kinks or merge this PR without unrolling.

**Comment by ggerganov** [comment]
> I think I've found an optimization option that's 5% faster than loop unrolling on my RTX 3090 and where loop unrolling actually degrades performance. It essentially works by processing more values per iteration in the loop instead of unrolling the loop. Implementation could be a little tricky. So you can either wait until I've worked out the kinks or merge this PR without unrolling.

Ok, I have mixed feelings about this unrolling anyway, so an alternative would be welcome

**Comment by JohannesGaessler** [comment]
I ended up doing an implementation kind of similar to what ggerganov proposed. It's now possible to set an option `LLAMA_CUDA_BX` that sets the how many data values in x direction are processed in each outer loop iteration. This then also increases the amount of data processed in the unrolled inner loop. The optimal parameters for my RTX 3090 are `LLAMA_CUDA_BX=64 LLAMA_CUDA_BY=2` which result in 14.53 t/s for 33b q4_0.

In terms of features I think this PR is complete but I just noticed that "BLOCK_X" is maybe a bad name for this value since it does not actually set the block size; that value is always `WARP SIZE` = 32. Suggestions for better naming?

**Comment by JohannesGaessler** [comment]
Maybe just `CUDA_DMMV_X` / `CUDA_DMMV_Y`?

**Review by ggerganov — APPROVED** [review]
Looks great. Will do some tests later

**Review by ggerganov — APPROVED** [review]
Minor fixes in Makefile

Works fine on 4080. `LLAMA_CUDA_DMMV_X=64 LLAMA_CUDA_DMMV_Y=1` seems to be optimal on this card, but haven't done extensive testing

**Comment by JohannesGaessler** [comment]
From my side yes.

**Comment by KerfuffleV2** [comment]
(Sorry about that, I accidentally hit a random key with the page focused.)

**Comment by SlyEcho** [comment]
This is amazing now, <70 ms/t for 13b Q4_0 on my old graphics card.

The variables need to be tuned though for different systems.

**Pull request merged**
